File: community/cypher/frontend-3.3/src/test/scala/org/neo4j/cypher/internal/frontend/v3_3/ast/functions/PointTest.scala
/*
 * Copyright (c) 2002-2017 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.cypher.internal.frontend.v3_3.ast.functions

import org.neo4j.cypher.internal.frontend.v3_3.symbols._

class PointTest extends FunctionTestBase("point")  {

  test("should accept correct types") {
    testValidTypes(CTMap)(CTPoint)
  }

  test("should fail type check for incompatible arguments") {
    // TODO: Fix error message which seems to assume that Node and Relationships are acceptable maps
    testInvalidApplication(CTList(CTAny))(
      "Type mismatch: expected Map, Node or Relationship but was List<Any>"
    )
    testInvalidApplication(CTString)(
      "Type mismatch: expected Map, Node or Relationship but was String"
    )
  }

  test("should fail if wrong number of arguments") {
    testInvalidApplication()(
      "Insufficient parameters for function 'point'"
    )
    testInvalidApplication(CTMap, CTMap)(
      "Too many parameters for function 'point'"
    )
  }
}

File: community/values/src/test/java/org/neo4j/values/virtual/PointTest.java
/*
 * Copyright (c) 2002-2017 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.values.virtual;

import org.junit.Test;

import static org.junit.Assert.assertTrue;
import static org.neo4j.values.virtual.VirtualValueTestUtil.assertEqual;
import static org.neo4j.values.virtual.VirtualValueTestUtil.assertNotEqual;
import static org.neo4j.values.virtual.VirtualValues.pointCartesian;
import static org.neo4j.values.virtual.VirtualValues.pointGeographic;

public class PointTest
{
    @Test
    public void cartesianShouldEqualItself()
    {
        assertEqual( pointCartesian( 1.0, 2.0 ), pointCartesian( 1.0, 2.0 ) );
        assertEqual( pointCartesian( -1.0, 2.0 ), pointCartesian( -1.0, 2.0 ) );
        assertEqual( pointCartesian( -1.0, -2.0 ), pointCartesian( -1.0, -2.0 ) );
        assertEqual( pointCartesian( 0.0, 0.0 ), pointCartesian( 0.0, 0.0 ) );
    }

    @Test
    public void cartesianShouldNotEqualOtherPoint()
    {
        assertNotEqual( pointCartesian( 1.0, 2.0 ), pointCartesian( 3.0, 4.0 ) );
        assertNotEqual( pointCartesian( 1.0, 2.0 ), pointCartesian( -1.0, 2.0 ) );
    }

    @Test
    public void geographicShouldEqualItself()
    {
        assertEqual( pointGeographic( 1.0, 2.0 ), pointGeographic( 1.0, 2.0 ) );
        assertEqual( pointGeographic( -1.0, 2.0 ), pointGeographic( -1.0, 2.0 ) );
        assertEqual( pointGeographic( -1.0, -2.0 ), pointGeographic( -1.0, -2.0 ) );
        assertEqual( pointGeographic( 0.0, 0.0 ), pointGeographic( 0.0, 0.0 ) );
    }

    @Test
    public void geographicShouldNotEqualOtherPoint()
    {
        assertNotEqual( pointGeographic( 1.0, 2.0 ), pointGeographic( 3.0, 4.0 ) );
        assertNotEqual( pointGeographic( 1.0, 2.0 ), pointGeographic( -1.0, 2.0 ) );
    }

    @Test
    public void geographicShouldNotEqualCartesian()
    {
        assertNotEqual( pointGeographic( 1.0, 2.0 ), pointCartesian( 1.0, 2.0 ) );
    }

    @Test
    public void shouldHaveValueGroup()
    {
        assertTrue( pointCartesian( 1, 2 ).valueGroup() != null );
        assertTrue( pointGeographic( 1, 2 ).valueGroup() != null );
    }
}
